app/src/main/java/com/naman14/timberx/ui/dialogs/CreatePlaylistDialog.kt

```diff
@@ -16,63 +16,63 @@ package com.naman14.timberx.ui.dialogs
-        return MaterialDialog(activity!!).show {
-            title(text = "Create new playlist")
-            positiveButton(text = "Create")
-            negativeButton(text = "Cancel")
-            input(hint = "Enter playlist name", callback = { dialog, text ->
-                val songs = arguments?.getLongArray("songs")
-
-                val playistId = MusicUtils.createPlaylist(activity!!, text.toString())
-
-                if (playistId.toInt() != -1) {
-                    if (songs != null && songs.isNotEmpty())
-                        MusicUtils.addToPlaylist(activity!!, songs, playistId)
-                    else
-                        Toast.makeText(activity, "Created playlist", Toast.LENGTH_SHORT).show()
-                    callback()
-                } else {
-                    Toast.makeText(activity, "Unable to create playlist", Toast.LENGTH_SHORT).show()
-                }
-            })
-        }
+    interface PlaylistCreatedCallback {
+        fun onPlaylistCreated()
     }
 
     companion object {
-        @JvmOverloads
-        fun newInstance(song: Song? = null): CreatePlaylistDialog {
+        private const val TAG = "CreatePlaylistDialog"
+
+        fun <T> show(parent: T, song: Song? = null) where T : Fragment, T : PlaylistCreatedCallback {
             val songs: LongArray
             if (song == null) {
                 songs = LongArray(0)
             } else {
                 songs = LongArray(1)
                 songs[0] = song.id
             }
-            return newInstance(songs)
+            show(parent, songs)
         }
 
-        fun newInstance(songList: LongArray): CreatePlaylistDialog {
-            return CreatePlaylistDialog().apply {
+        fun <T> show(parent: T, songList: LongArray) where T : Fragment, T : PlaylistCreatedCallback {
+            val dialog = CreatePlaylistDialog().apply {
                 arguments = Bundle().apply { putLongArray("songs", songList) }
             }
+            dialog.show(parent.childFragmentManager, TAG)
+        }
+    }
+
+    @NonNull
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        return MaterialDialog(activity!!).show {
+            title(text = "Create new playlist")
+            positiveButton(text = "Create")
+            negativeButton(text = "Cancel")
+            input(hint = "Enter playlist name", callback = { _, text ->
+                val songs = arguments?.getLongArray("songs")
+                val playlistId = MusicUtils.createPlaylist(activity!!, text.toString())
+
+                if (playlistId.toInt() != -1) {
+                    if (songs != null && songs.isNotEmpty()) {
+                        MusicUtils.addToPlaylist(activity!!, songs, playlistId)
+                    } else {
+                        activity.toast("Created playlist")
+                    }
+                    (parentFragment as? PlaylistCreatedCallback)?.onPlaylistCreated()
```

Comment:
This is safe for screen rotations because the parent fragment itself is a `PlaylistCreatedCallback` rather than having to rely on a non-persistent callback instance 😀

Reply:
But does this work across process death? :thinking: 

I'm sure `setTargetFragment` works because it's magical.

Reply:
@Zhuinden it does indeed - but yeah, `setTargetFragment` could work too

Reply:
Cool trick!